discovery-provider/integration_tests/tasks/test_index_trending_underground_notification.py

```diff
@@ -0,0 +1,182 @@
+        set_json_cached_key(redis, trending_key, trending_underground_tracks)
+
+
+def test_index_trending_underground_notification(app, caplog):
+    """
+    Test that underground notifications are correctly calculated
+    based on current trending standings
+    """
+    caplog.set_level(logging.WARNING)
+
+    with app.app_context():
+        db = get_db()
+        # Add some users to the db so we have blocks
+        entities = {
+            "tracks": [
+                {
+                    "track_id": i,
+                    "owner_id": i % 2 + 1,
+                }
+                for i in range(100)
+            ],
+            "users": [{}] * 3,
+        }
+        populate_mock_db(db, entities)
+
+        # Test that if there are no prev notifications, all trending in top are generated
+        index_underground_trending_mock(db, list(range(1, 21)))
+        base_time_int = int(round(BASE_TIME.timestamp()))
+        index_trending_underground_notifications(db, base_time_int)
+
+        with db.scoped_session() as session:
+            notifications = (
+                session.query(Notification).order_by(asc(Notification.specifier)).all()
+            )
+            assert len(notifications) == 5
+
+            notification_0 = notifications[0]
+            assert notification_0.specifier == "1"
+            assert (
+                notification_0.group_id
+                == "trending_underground:time_range:week:genre:all:rank:1:track_id:1:timestamp:1672531200"
+            )
+            assert notification_0.type == "trending_underground"
+            assert notification_0.data == {
+                "rank": 1,
+                "genre": "all",
+                "track_id": 1,
+                "time_range": "week",
+            }
+            assert notification_0.user_ids == [2]
+
+            notification_1 = notifications[1]
+            assert notification_1.specifier == "2"
+            assert (
+                notification_1.group_id
+                == "trending_underground:time_range:week:genre:all:rank:2:track_id:2:timestamp:1672531200"
+            )
+            assert notification_1.type == "trending_underground"
+            assert notification_1.data == {
+                "rank": 2,
+                "genre": "all",
+                "track_id": 2,
+                "time_range": "week",
+            }
+            assert notification_1.user_ids == [1]
+
+        # Test that on second iteration of trending, only new notifications appear
+        # Prev: 1, 2, 3, 4, 5
+        # New: 2, 4, 5, 1, 7
+        # new notifications for track_id: 2, 4, 5, 7
+        # no new notification for track_id: 1
+
+        index_underground_trending_mock(db, [2, 4, 5, 1, 7, 8, 9, 10])
+        updated_time = BASE_TIME + timedelta(hours=1)
+        updated_time_int = int(round(updated_time.timestamp()))
+        index_trending_underground_notifications(db, updated_time_int)
```

Comment:
for my learning could you explain this test case a bit more? how does it choose which notif to update?

Reply:
yeah i pretty much copied it from the original trending test. so what this doing, is saying now that we have sent out notifs for tracks 1-5, if we reindex trending again and see that 2,4,5,1,7 are the new top 5, we should send out notifications for 2,4,5, and 7, since each one except for 1 increased their trending number.